### DISCUSSION TITLE
Where the init of font I can found in the code?

### DISCUSSION THREAD

**Discussion by utt54321** [tsujan/FeatherPad#655] [Q&A]
I was reading this project's code. I found in the config.cpp it sets font as "Monospace".
<img width="187" alt="image" src="https://user-images.githubusercontent.com/20319237/170860837-c9aae36e-2bef-4bcd-8425-66b1150a3cb4.png">
But I compile it, and install it and make sure that there is no .conf file in my ~/.config/featherpad, then start running featherpad, the font is default "Noto Sans Mono CJK SC"
<img width="351" alt="image" src="https://user-images.githubusercontent.com/20319237/170860991-5f4c3115-ec7d-4eb2-92d3-461b35763828.png">
I cannot find out where the default font the code sets. Can someone helps telling the story of the initial process of font with showing code?

**Comment by utt54321** [comment]
Anyone help me please, >~<

**Comment by tsujan** [comment]
> the font is default "Noto Sans Mono CJK SC"

When FeatherPad is started for the first time, the font is `QFont("Monospace")` and its size is that of the app's font size or 9, whichever is greater.

The system is responsible for providing a good Monospace font; if the system doesn't do it correctly, the user could change the font.

The relevant code lines are https://github.com/tsujan/FeatherPad/blob/cf4d668975567b1a6fe3fb738e690e3464adbab8/featherpad/config.cpp#L134 and https://github.com/tsujan/FeatherPad/blob/cf4d668975567b1a6fe3fb738e690e3464adbab8/featherpad/config.cpp#L255-L259

**Reply by utt54321** [reply]
You mean "Noto Sans Mono CJK SC" is one of types of "Monospace"? So when the system dose not have "Monospace", featherpad will choose a proper font. But how dose featherpad choosing font? In my another system, its font at fist running is "UKIJ CJK":
 
![image](https://user-images.githubusercontent.com/20319237/170862044-06ab3a08-9db8-410b-804b-d1a2cdcf081f.png)

**Reply by tsujan** [reply]
> You mean "Noto Sans Mono CJK SC" is one of types of "Monospace"?

Yes, as its name says, it's a Monospace font. For whatever reason, it's the default Monospace font in your system.

> So when the system dose not have "Monospace", featherpad will choose a proper font...

Actually, Qt does it; FeatherPad just uses what Qt does. And Qt does it properly. As mentioned above, the system *has* "Monospace" -- it's "Noto Sans Mono CJK SC" in your case.

Please read my other comment about changing Monospace (if you want to change it). After changing it, close *all* FeatherPad windows and open a window again. It isn't limited to FeatherPad, of course.

**Reply by tsujan** [reply]
BTW, if you want to know what the default Monospace font is in your system, you could use this command:

```
fc-match monospace
```

**Reply by utt54321** [reply]
Is there any website or pdf shows the info like "Noto Sans Mono CJK SC belongs to Monospace"? So I can find out why in the another system the font can be "UKIJ CJK".

**Reply by utt54321** [reply]
fc-match monospace helps a lot !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
Very very very very thank you, I wanna hug you face to face.

**Reply by tsujan** [reply]
Fonts that have "mono" in their names are monospace; otherwise, they should have a bug.

If you check "Programming Fonts" in the "Select Font" dialog (which is checked by default), you'll see all monospace fonts installed on your system. If you uncheck it, you'll see all fonts, whether monospace or not.

**Reply by tsujan** [reply]
You may already know it but "monospace" is the property that some fonts have: the widths of all characters are the same in them: "I" and "m" have the same width with them, for example. That makes them ideal for coding and similar jobs.

**Comment by tsujan** [comment]
If your question was really about a bad Monospace font in your system, you could change it by creating `~/.config/fontconfig/conf.d/0-mono.conf` as (replace "DejaVu" with any font you may prefer):

```xml
<?xml version='1.0'?>
<!DOCTYPE fontconfig SYSTEM 'fonts.dtd'>
<fontconfig>
 <alias>
  <family>monospace</family>
  <prefer>
   <family>DejaVu Sans Mono</family>
  </prefer>
 </alias>
 <alias>
  <family>serif</family>
  <prefer>
   <family>DejaVu Serif</family>
  </prefer>
 </alias>
 <alias>
  <family>sans-serif</family>
  <prefer>
   <family>DejaVu Sans</family>
  </prefer>
 </alias>
</fontconfig>

```
The above example covers "serif" and "sans-serif" in addition to "monospace", for the sake of completeness.

**Comment by utt54321** [comment]
How to make featherpad follow the system's font at any time?
Or how to code [FeatherPad/featherpad/config.cpp] to make the default font to be the system's font?

**Reply by tsujan** [reply]
As a code editor, FeatherPad's view should use the system's *monospace* font by default, not the system font. Of course, the user could choose any font later.

Other parts of the window follow the system font, as seen by Qt (which may be different from what GTK uses).

**Reply by utt54321** [reply]
I am just interesting about how to get the system's font in the  [FeatherPad/featherpad/config.cpp] and then set the system's font as the default font (replace monospace).
But I am not very good at QT (I have learned about it only one hour by searching on Internet), so I can not know how to code it.

**Reply by tsujan** [reply]
If you want the system font in your Qt app (which is good practice for all apps other than text editors), just don't set the font in your code. If no font is set, the default font of the system will be used.

Qt requires platform integration. Qt-based desktop environments like KDE and LXQt provide it automatically. With other DEs, you need a Qt config tool (like qt5ct/qt6ct).

Qt follows `freedesktop.org` standards. As for font setting, `freedesktop.org` has a comprehensive documentation; see https://www.freedesktop.org/software/fontconfig/fontconfig-user.html

**Reply by tsujan** [reply]
As for FeatherPad, if you uncheck `Preferences  → Text → Remember font`, the system's default monospace font will be followed. Then, the "Select Font" dialog will set the font only temporarily, i.e., only for the current page/tab.

**Reply by utt54321** [reply]
I have read the https://www.freedesktop.org/software/fontconfig/fontconfig-user.html.
But I still have no idea how to use QT read the system's font, how QT dose it.
And what is the meaning of the checkbox "Programming Font", it seems no defference between it sets and not.

**Reply by tsujan** [reply]
> But I still have no idea how to use QT read the system's font

As I said earlier, you don't need to do anything. The c-tor of `QFont` does it.

> And what is the meaning of the checkbox "Programming Font"

Programming Fonts =  Coding Fonts = Monospace Fonts. It makes a big difference: it determines what's shown by the drop-down list below it.

**Reply by utt54321** [reply]
So sorry. I am too stupid to understand the QFont you mentioned.
Do you mean QFont itself has a variable of the system's font? But I did not find where it shows the system's font.

**Reply by tsujan** [reply]
It's just a matter of getting familiar with Qt :) You'll master it after a while.

> Do you mean QFont itself has a variable of the system's font?

I mean, if you write,
```c++
auto f = QFont();
```
Then `f` is the font that your system uses. You don't need anything else to find it; Qt does it for you.

**Reply by utt54321** [reply]
Oh my god!!!!!!!
I am crying for its simplity. Thank you.
Sir, how you learn QT so well, can I copy your learning path? And can I get your email?

**Reply by tsujan** [reply]
Ha ha, yes, Qt makes many things easy :)

My email is inside [README](https://github.com/tsujan/FeatherPad/blob/master/README.md).
